cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java

```diff
@@ -0,0 +1,302 @@
+          }
+        });
+
+    // Subscriber fields
+    this.kafkaCallbacks = Maps.newConcurrentMap();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    publishingExecutor = MoreExecutors.listeningDecorator(
+      Executors.newSingleThreadExecutor(Threads.createDaemonThreadFactory("notification-publisher-%d")));
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    publishingExecutor.shutdownNow();
+  }
+
+  @Override
+  public <N> ListenableFuture<N> publish(NotificationFeed feed, N notification)
+    throws NotificationException {
+    return publish(feed, notification, notification.getClass());
+  }
+
+  @Override
+  public <N> ListenableFuture<N> publish(final NotificationFeed feed, final N notification, final Type notificationType)
+    throws NotificationException {
+    LOG.debug("Publishing on notification feed [{}]: {}", feed, notification);
+    return publishingExecutor.submit(new Callable<N>() {
+      @Override
+      public N call() throws Exception {
+        try {
+          KafkaMessage message = new KafkaMessage(KafkaNotificationUtils.getMessageKey(feed),
+                                                  GSON.toJsonTree(notification, notificationType));
+          ByteBuffer bb = KafkaMessageCodec.encode(message);
+
+          TopicPartition topicPartition = KafkaNotificationUtils.getKafkaTopicPartition(feed);
+          KafkaPublisher.Preparer preparer = kafkaPublishersCache.get(topicPartition.getTopic());
+          preparer.add(bb, message.getMessageKey());
+
+          try {
+            preparer.send().get();
+            return notification;
+          } catch (ExecutionException e) {
+            throw new NotificationException(e.getCause());
+          }
+        } catch (IOException e) {
+          throw new NotificationException(e);
+        }
+      }
+    });
+  }
+
+  private KafkaPublisher getKafkaPublisher() {
+    if (kafkaPublisher != null) {
+      return kafkaPublisher;
+    }
+    try {
+      kafkaPublisher = kafkaClient.getPublisher(ack, Compression.SNAPPY);
+    } catch (IllegalStateException e) {
+      // can happen if there are no kafka brokers because the kafka server is down.
+      kafkaPublisher = null;
+    }
+    return kafkaPublisher;
+  }
+
+
+  @Override
+  public <N> Cancellable subscribe(NotificationFeed feed, NotificationHandler<N> handler)
+    throws NotificationFeedException {
+    return subscribe(feed, handler, Threads.SAME_THREAD_EXECUTOR);
+  }
+
+  @Override
+  public <N> Cancellable subscribe(final NotificationFeed feed, final NotificationHandler<N> handler,
+                                   Executor executor) throws NotificationFeedException {
+    // This call will make sure that the feed exists
+    feedManager.getFeed(feed);
+
+    final TopicPartition topicPartition = KafkaNotificationUtils.getKafkaTopicPartition(feed);
+
+    KafkaNotificationsCallback kafkaNotificationsCallback;
+    synchronized (kafkaCallbacks) {
```

Comment:
Because I want the whole block to be synced, not just the calls to the map. If two threads arrive here, and there wasn't this synchronized block, there is a chance that the two threads would create two different ``KafkaNotificationsCallback`` objects for a same topic partition.